### PR TITLE
[Spriest] Chorus of Insanity azerite trait

### PR DESCRIPTION
**Pull request by TJMoats**
screen shot 2018-11-07 at 9 07 23 am

### CODE REVIEW

File: src/parser/priest/shadow/modules/spells/azeritetraits/ChorusOfInsanity.js [hunk 1]
```diff
@@ -0,0 +1,123 @@
...
+      // This check adds the first tick.
+      if (this.pendingStack) {
+        this.totalCrit += (this.currentStacks + 1) * this.crit * uptimeOnStack;
+        this.pendingStack = false;
+      }
+
+      this.totalCrit += this.currentStacks * this.crit * uptimeOnStack;
+    }
+
+    if (event.type === "applybuff") {
```
Comment (Dambroda) [inline src/parser/priest/shadow/modules/spells/azeritetraits/ChorusOfInsanity.js]: Small one but just for style convention should be single quotes (and removebuff as well)

File: src/parser/priest/shadow/modules/spells/azeritetraits/ChorusOfInsanity.js [hunk 2]
```diff
@@ -0,0 +1,123 @@
...
+
+    this.lastTimestamp = event.timestamp;
+  }
+
+  get uptime() {
+    return this.selectedCombatant.getBuffUptime(SPELLS.CHORUS_OF_INSANITY_BUFF.id) / this.owner.fightDuration;
+  }
+
+  get averageCrit() {
+    return (this.totalCrit / this.owner.fightDuration).toFixed(0);
```
Comment (Dambroda) [inline src/parser/priest/shadow/modules/spells/azeritetraits/ChorusOfInsanity.js]: Couldn't you use _getStackWeightedBuffUptime_ to pretty much completely avoid having to do any of these calculations yourself?

File: src/parser/priest/shadow/modules/spells/azeritetraits/ChorusOfInsanity.js [hunk 3]
```diff
@@ -0,0 +1,123 @@
...
+
+  statistic() {
+    return (
+      <TraitStatisticBox
+        position={STATISTIC_ORDER.OPTIONAL()}
+        trait={SPELLS.CHORUS_OF_INSANITY.id}
+        value={`${this.averageCrit} average Crit`}
+        tooltip={`
+          ${SPELLS.CHORUS_OF_INSANITY.name} grants <b>${this.crit} crit per stack</b><br/>
+          You procced <b>${SPELLS.CHORUS_OF_INSANITY.name} ${this.chorusProcs} times</b> with an uptime of ${formatPercentage(this.uptime)}%.
```
Comment (Dambroda) [inline src/parser/priest/shadow/modules/spells/azeritetraits/ChorusOfInsanity.js]: It's only stylistic, but arent the numbers usually bolded and the names not (only in second line here):

Currently:
You procced **Corus of Insanity X times** with an uptime of X%.

Standard (imo):
You procced Corus of Insanity **X times** with an uptime of **X%**.

File: src/parser/priest/shadow/modules/spells/azeritetraits/ChorusOfInsanity.js (hunk 2 above)
Comment (TJMoats) [reply src/parser/priest/shadow/modules/spells/azeritetraits/ChorusOfInsanity.js]: getStackWeightedBuffUptime doesn't calculate the initial stack application correctly. There is no initial applybuffstack event, so I need the custom logic that I have here.

File: src/parser/priest/shadow/modules/spells/azeritetraits/ChorusOfInsanity.js (hunk 3 above)
Comment (TJMoats) [reply src/parser/priest/shadow/modules/spells/azeritetraits/ChorusOfInsanity.js]: Fixed

File: src/parser/priest/shadow/modules/spells/azeritetraits/ChorusOfInsanity.js (hunk 1 above)
Comment (TJMoats) [reply src/parser/priest/shadow/modules/spells/azeritetraits/ChorusOfInsanity.js]: Fixed

File: src/parser/priest/shadow/modules/spells/azeritetraits/ChorusOfInsanity.js (hunk 2 above)
Comment (Dambroda) [reply src/parser/priest/shadow/modules/spells/azeritetraits/ChorusOfInsanity.js]: Wouldn't it be better to fabricate the buff update event in that case?

Comment (TJMoats) [reply src/parser/priest/shadow/modules/spells/azeritetraits/ChorusOfInsanity.js]: Do you have an example of how to do that? and I don't have the number of stacks until the first removebuffstack. Would I be able to fabricate an event a second before the removebuffstack event?

Comment (Dambroda) [reply src/parser/priest/shadow/modules/spells/azeritetraits/ChorusOfInsanity.js]: Oh, I think in this case you could call something like:
```javascript
    this.owner.fabricateEvent({
      ...event, // copies everything from the removebuffstack event except below
      type: 'applybuffstack',
      stack: event.stack + 1,      
      timestamp: /*timestamp copied from the applybuff event; maybe add 1ms?*/,
    }, event);
```

And I'd keep your existing math in there as well to test and ensure you get the same results from both methods as well

Comment (TJMoats) [reply src/parser/priest/shadow/modules/spells/azeritetraits/ChorusOfInsanity.js]: I think I may have fabricated the event properly. The numbers are within margin of error.

Comment (Dambroda) [reply src/parser/priest/shadow/modules/spells/azeritetraits/ChorusOfInsanity.js]: It looks like you may be missing the source/targets ids which may lead to the non-matching numbers

Comment (TJMoats) [reply src/parser/priest/shadow/modules/spells/azeritetraits/ChorusOfInsanity.js]: It was accurate to within .5 crit of what I had for the final number. Good enough for me

Comment (Dambroda) [reply src/parser/priest/shadow/modules/spells/azeritetraits/ChorusOfInsanity.js]: Could you add the ids? They are needed for pretty much all all event filtering I think

Comment (TJMoats) [reply src/parser/priest/shadow/modules/spells/azeritetraits/ChorusOfInsanity.js]: I just added the event

**Pull request merged**